api/krusty/remoteloader_test.go

```diff
@@ -0,0 +1,348 @@
+	// root/
+	//   simple.git/			- base with just a pod
+	//   multibase.git/			- base with a dev overlay
+	//   with-submodule.git/	- includes `simple` as a submodule
+	//     submodule/			- the submodule referencing `simple
+	createGitRepos := func(t *testing.T) testRepos {
+		t.Helper()
+
+		bash := func(script string) {
+			cmd := exec.Command("sh", "-c", script)
+			o, err := cmd.CombinedOutput()
+			if err != nil {
+				t.Fatalf("error running %v\nerr: %v\n%s", script, err, string(o))
+			}
+		}
+		root := t.TempDir()
+		bash(fmt.Sprintf(`
+export ROOT="%s"
+set -eux
+cp -r testdata/remoteload/simple $ROOT/simple.git
+(
+	cd $ROOT/simple.git
+	git init --initial-branch=main
+	git add .
+	git commit -m "import"
+	git checkout -b change-image
+	cat >>kustomization.yaml <<EOF
+
+images:
+- name: nginx
+  newName: nginx
+  newTag: "2"
+EOF
+	git commit -am "image change"
+	git checkout main
+)
+cp -r testdata/remoteload/multibase $ROOT/multibase.git
+(
+	cd $ROOT/multibase.git
+	git init --initial-branch=main
+	git add .
+	git commit -m "import"
+)
+(
+	mkdir $ROOT/with-submodule.git
+	cd $ROOT/with-submodule.git
+	git init --initial-branch=main
+	git submodule add $ROOT/simple.git submodule
+	git commit -m "import"
+)
+`, root))
+		return testRepos{
+			root:          root,
+			simple:        "simple.git",
+			multiBaseDev:  "multibase.git",
+			withSubmodule: "with-submodule.git",
+		}
+	}
+
+	const simpleBuild = `apiVersion: v1
+kind: Pod
+metadata:
+  labels:
+    app: myapp
+  name: myapp-pod
+spec:
+  containers:
+  - image: nginx:1.7.9
+    name: nginx
+`
+	var multibaseDevExampleBuild = strings.ReplaceAll(simpleBuild, "myapp-pod", "dev-myapp-pod")
+
+	repos := createGitRepos(t)
+	tests := []struct {
+		name          string
+		kustomization string
+		expected      string
+		err           string
+		skip          bool
+	}{
+		{
+			name: "simple",
+			kustomization: `
+resources:
+- file://$ROOT/simple.git
+`,
+			expected: simpleBuild,
+		},
+		{
+			name: "has path",
+			kustomization: `
+resources:
+- file://$ROOT/multibase.git/dev
+`,
+			expected: multibaseDevExampleBuild,
+		},
+		{
+			name: "has ref",
+			kustomization: `
+resources: 
+- "file://$ROOT/simple.git?ref=change-image"
+`,
+
+			expected: strings.ReplaceAll(simpleBuild, "nginx:1.7.9", "nginx:2"),
+		},
+		{
+			// Version is the same as ref
+			name: "has version",
+			kustomization: `
+resources:
+- file://$ROOT/simple.git?version=change-image
+`,
+			expected: strings.ReplaceAll(simpleBuild, "nginx:1.7.9", "nginx:2"),
+		},
+		{
+			name: "has submodule",
+			kustomization: `
+resources:
+- file://$ROOT/with-submodule.git/submodule
+`,
+			expected: simpleBuild,
+		},
+		{
+			name: "has submodule but not initialized",
+			kustomization: `
+resources:
+- file://$ROOT/with-submodule.git/submodule?submodules=0
+`,
+			err: "unable to find",
+		},
+		{
+			name: "has origin annotation",
+			skip: true, // The annotated path should be "pod.yaml" but is "notCloned/pod.yaml"
```

Comment:
Calling out that there may be an existing bug.

Reply:
Looked into this briefly, and I think it is due to the special handling of the file protocol specifically. The path given to `Origin#Append` is not parsing as a RepoSpec, I think because it does not have `.git` in it. The error is `url lacks orgRepo: pod.yaml`, and the fallback essentially assumes that the origin in question is not a clone. So I think the problem is new to this PR, and another reason to make the file protocol handling as similar as possible to the others.

Reply:
Ack, you are right. This goes back to the problem where `GitSuffix` has some special treatment that doesn't work well for the `file://` scheme. I made some changes that try to make it play a bit nicer.

Reply:
Can you now enable this test? Since the problem is novel to this PR, I would expect it to be working after your changes.

Reply:
It appears that this is not novel to this PR. I've created https://github.com/kubernetes-sigs/kustomize/pull/4801 as a minimal repro case on the existing code.